histogram/src/sparse.rs

```diff
@@ -0,0 +1,304 @@
+    pub count: Vec<u64>,
+}
+
+impl SparseHistogram {
+    fn add_bucket(&mut self, idx: usize, n: u64) {
+        if n != 0 {
+            self.index.push(idx);
+            self.count.push(n);
+        }
+    }
+
+    /// Merges two Histograms and returns the results in a new Histogram.
+    ///
+    /// Both histograms must have the same configuration parameters.
+    /// Buckets which have values in both histograms are allowed to wrap.
+    #[allow(clippy::comparison_chain)]
+    pub fn merge(&self, h: &SparseHistogram) -> Result<SparseHistogram, Error> {
+        if self.config != h.config {
+            return Err(Error::IncompatibleParameters);
+        }
+
+        let mut histogram = SparseHistogram {
+            config: self.config,
+            index: Vec::new(),
+            count: Vec::new(),
+        };
+
+        // Sort and merge buckets from both histograms
+        let (mut i, mut j) = (0, 0);
+        while i < self.index.len() && j < h.index.len() {
+            let (k1, v1) = (self.index[i], self.count[i]);
+            let (k2, v2) = (h.index[j], h.count[j]);
+
+            if k1 == k2 {
+                histogram.add_bucket(k1, v1 + v2);
+                (i, j) = (i + 1, j + 1);
+            } else if k1 < k2 {
+                histogram.add_bucket(k1, v1);
+                i += 1;
+            } else {
+                histogram.add_bucket(k2, v2);
+                j += 1;
+            }
+        }
+
+        // Fill remaining values, if any, from the left histogram
+        if i < self.index.len() {
+            histogram.index.extend(&self.index[i..self.index.len()]);
+            histogram.count.extend(&self.count[i..self.count.len()]);
+        }
+
+        // Fill remaining values, if any, from the right histogram
+        if j < h.index.len() {
+            histogram.index.extend(&h.index[i..h.index.len()]);
+            histogram.count.extend(&h.count[i..h.count.len()]);
+        }
+
+        Ok(histogram)
+    }
+
+    /// Return a single percentile from this histogram.
+    ///
+    /// The percentile should be in the inclusive range `0.0..=100.0`. For
+    /// example, the 50th percentile (median) can be found using `50.0`.
+    pub fn percentile(&self, percentile: f64) -> Result<Bucket, Error> {
+        let total: u128 = self.count.iter().map(|v| *v as u128).sum();
+
+        if !(0.0..=100.0).contains(&percentile) {
+            return Err(Error::InvalidPercentile);
+        }
+
+        let search = ((total as f64) * percentile / 100.0).ceil() as usize;
+        let mut seen: usize = 0;
+        for (idx, count) in self.index.iter().zip(self.count.iter()) {
+            seen += *count as usize;
+            if seen >= search {
+                return Ok(Bucket {
+                    count: *count,
+                    range: self.config.index_to_range(*idx),
+                });
+            }
+        }
+
+        // should be unreachable
+        Err(Error::Unreachable)
```

Comment:
Can we do something different here? If we're never going to hit this (which we really shouldn't) I'd consider either panic or returning the highest bucket with a count. Since this is a sparse representation, that's just the last bucket.

Reply:
Fixed with last bucket.